### DISCUSSION TITLE
Still having problems with Interactions.s.sol but the script itself runs successfully...getting errors on Sepolia, Anvil, and forked Sepolia

### DISCUSSION THREAD

**Discussion by 0xwtfwtf** [Cyfrin/foundry-full-course-cu#552] [General]
I had so many problems getting Interactions.s.sol and Interactions.t.sol working but I finally got it semi-working such that the tests all passed on Sepolia. They don't pass on anvil though. So I decided to back up a step just run script FundFundMe from Interactions.s.sol to see if that worked on its own. I tried it on Sepolia and Anvil. On Sepolia, the script ran successfully and I got the proper console.log output to confirm that but I still got an internal EVM error (first error below). I looked this up and tried some of the suggested fixes like adding solc="0.8.19" to foundry.toml and installing foundryup but those didn't work. Next I tried running script FundFundMe from Interactions.s.sol on anvil and the script didn't run properly so I didn't get the right console.log output. I tried to first run forge script script/DeployFundMe.sol on anvil to see if it just needed the contract to be deployed first but that didn't work.

Edit: [I posted my code on GitHub](github.com/0xwtfwtf/foundry-f23-FundMe)

Error when running "forge script script/Interactions.s.sol:FundFundMe --fork-url $SEPOLIA_PRIVATE_KEY (get the same error when I run it on the actual Sepolia network instead of forking):

> [⠘] Compiling...
No files changed, compilation skipped
Script ran successfully.

== Logs ==
  Return Data:
  0xf78e4178a013174a414482857a9c4a95ab70e589
  Funded FundMe with 1000000000000000000

**## Setting up (1) EVMs.
The application panicked (crashed).
Message:  Internal EVM error: 
fuzz: failed to inspect: Transaction(LackOfFundForGasLimit { gas_limit: 0x0000000000000000000000000000000000000000000000000de0b6b3a7640000_U256, balance: 0x0000000000000000000000000000000000000000000000000000000000000000_U256 })
Location: /home/runner/work/foundry/foundry/cli/src/cmd/forge/script/executor.rs:158

This is a bug. Consider reporting it at https://github.com/foundry-rs/foundry**

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
Aborted <

Error when running "forge script script/Interactions.s.sol:FundFundMe" on anvil:

> forge script script/Interactions.s.sol:FundFundMe
[⠑] Compiling...
No files changed, compilation skipped
Traces:
  [639098] → new DevOpsTools@0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496
    └─ ← 3192 bytes of code

  [247908] → new FundFundMe@0x34A1D3fff3958843C43aD80F30b94c510645C316
    └─ ← 1017 bytes of code

  [59778] FundFundMe::run() 
    ├─ [35766] DevOpsTools::e374cdf1(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000007a69000000000000000000000000000000000000000000000000000000000000000646756e644d650000000000000000000000000000000000000000000000000000) [delegatecall]
    │   ├─ [0] VM::ffi([pwd]) 
    │   │   └─ ← 0x2f686f6d652f6563732f736f6c69646974792d636f757273652f666f756e6472792d6632332f666f756e6472792d66756e642d6d652d663233
    │   ├─ [0] VM::toString(31337 [3.133e4]) [staticcall]
    │   │   └─ ← 0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000053331333337000000000000000000000000000000000000000000000000000000
    │   ├─ [0] VM::ffi([bash, /home/ecs/solidity-course/foundry-f23/foundry-fund-me-f23/lib/foundry-devops/src/get_recent_deployment.sh, FundMe, 31337, /home/ecs/solidity-course/foundry-f23/foundry-fund-me-f23//broadcast]) 
    │   │   └─ ← 0x0000000000000000000000000000000000000000
    │   ├─ [0] console::log(Return Data:) [staticcall]
    │   │   └─ ← ()
    │   ├─ [0] console::log(0x0000000000000000000000000000000000000000) [staticcall]
    │   │   └─ ← ()
    │   └─ ← 0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000144e6f20636f6e7472616374206465706c6f796564000000000000000000000000
    └─ ← "No contract deployed"

Gas used: 59778

== Logs ==
  Return Data:
  0x0000000000000000000000000000000000000000
Error: 
No contract deployed>

**Comment by shikhar229169** [comment]
Can you provide the link for your repo. It would be much easier to debug it.

**Reply by 0xwtfwtf** [reply]
How do I upload my code to a repo? Ive never done that

**Reply by shikhar229169** [reply]
First make a new github repo: https://github.com/new
Add the name for your repo, add some description if you want and click create repository.

Then there will be a link something like: ```https://github.com/<your_name>/<your repo name>.git```
Copy that link.
Now go to the folder which you want to push on github. Ensure that .env file is in .gitignore.
Also add ```broadcast``` in .gitignore
Also ensure that you don't push any sensitive information like private key or rpc url.

Then in terminal run the command:
```
git remote add origin <the copied link>
git add .
git commit -m "any message"
git push origin master
```
Then paste the link of repo here,

**Reply by 0xwtfwtf** [reply]
Okay I did that and added you as collaborator...I first tried to use "publish to github" from vscode and it seemed to work but none of the files appeared. Then I used your method and I can see the files under "master" but they don't appear on the main page of the repo for some reason.

**Reply by 0xwtfwtf** [reply]
[I posted my code on GitHub](github.com/0xwtfwtf/foundry-f23-FundMe)

**Reply by shikhar229169** [reply]
I have tested your code on my system and it is working perfectly fine. I think that you are missing something while writing the commands.
Below are the steps you need to follow:


- First you need to deploy your smart contract to a network (say anvil). For this run the following command.
- Don't forget to do source .env
```
forge script script/DeployFundMe.s.sol --rpc-url $<name as in your env file> --private-key $<name as in your env file> --broadcast
```
- Then to run the Interactions script for funding
- Ensure that you have installed jq
```
forge script script/Interactions.s.sol:FundFundMe --rpc-url $<name as in your env file> --private-key $<name as in your env file> --broadcast
```

If it works try to do the same thing with sepolia rpc url and your account which has sepolia eth.

**Reply by 0xwtfwtf** [reply]
hmm that's so weird because those are the commands I am using...and I do have jq installed. Here is my screen cap showing what I get when I run the Interactions.s.sol second command exactly like what you said on anvil. And you can see I'm using the right terms for --rpc-url and --private-key since I have source .env open. And I ran the anvil command and source .env before running the script command too. When I run the Interactions.s.sol command on sepolia, the script runs successfully but I get an internal EVM error (I put the full error in my original post). And when I run forge test on anvil and sepolia, the tests pass which is even weirder because InteractionsTest.t.sol imports Interactions.s.sol so you would think it wouldn't work if Interactions.s.sol is throwing off errors.

Here is my screen cap for running the Interactions.s.sol script command on anvil:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/138697105/48898955-8953-4bfe-a9bf-db84dcc9dde8)

**Reply by shikhar229169** [reply]
Have you run the below command:
```
forge script script/DeployFundMe.s.sol --rpc-url $<name as in your env file> --private-key $<name as in your env file> --broadcast
```

It is necessary to run before running the interactions command.

**Comment by theirrationalone** [comment]
You need to install "jq"... And to set ffi = true in foundry.toml

Then follow this chronology... 

1. Run anvil chain in one terminal. 

2. Open another terminal and deploy main contract (FundMe??) On local anvil testnet.

Note: you can save deployed contract address for future use, if needed. 

3. Deploy interactions scripts... 

Now should work....

**Reply by 0xwtfwtf** [reply]
I already installed jq and I have ffi = true in foundry.toml. That was what enabled me to at least get the tests to pass on Sepolia. I also have the anvil RPC_URl and anvil PRIVATE_KEY in my source .env and i already have an anvil terminal opened and I tried running forge script script/Interactions.s.sol:FundFundMe --rpc-url $RPC_URL --private-key $PRIVATE_KEY --broadcast and I still get the same error. The error must have something to do with anvil since the script runs successfully on Sepolia (but throws off the other error about an internal evm error)

**Reply by theirrationalone** [reply]
To Test this test you would need to start anvil first,
then you'll have to deploy 'DeployScript' on anvil with broadcast
then you'll have to deploy 'FundFundMe' on anvil with broadcast and similarily
you'll have to deploy 'WithdrawFundMe' sequentially.
Test it when you use anvil or any other network explicitly -> command: forge test --rpc-url "any-url even anvil too" (✅)
Don't test it on implicit anvil -> command: forge test (❌)

```solidity
    // 👇 test function
    function testFundRunAndWithdrawRunFunctions() public {
        // Deployer(Owner) of FundMe is who Deployed FundMe (Hopefully DeployFundMe 'contract').

        // Deployer(Owner): self or FundFundMe is the owner also.
        // anybody can fund with eligible funding amount.
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.run();

        console.log("Current Owner: ", fundMe.getOwner());

        // Deployer(Owner): self or WithdrawFundMe is the owner also.
        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();

        // Can't withdraw funds because WithdrawFundMe isn't the Owner.
        // WithdrawFundMe is trying to withdraw funds from FundMe but Owner of FundMe is somebody else (hopefully DeployFundMe 'contract').
        vm.expectRevert();
        withdrawFundMe.run();
    }
```

**Comment by theirrationalone** [comment]
At first you ran for sepolia:

❌ here, You mistakenly put private key instead of rpc/fork url. 

```
forge script script/Interactions.s.sol:FundFundMe --fork-url $SEPOLIA_PRIVATE_KEY
```

It should be: ☑

```terminal
forge script script/Interactions.s.sol:FundFundMe --fork-url $YOUR_SEPOLIA_RPC/FORK_URL --private-key $YOUR_SEPOLIA_PRIVATE_KEY --broadcast
```

Then,
You ran for anvil:  This is wrong ❌ ... 

**Note:** Default ANVIL won't work. Specify rpc url amd private key explicitly. 

```terminal
forge script script/Interactions.s.sol:FundFundMe
```

It should be: ☑

```terminal
forge script script/Interactions.s.sol:FundFundMe --rpc-url $YOUR_ANVIL_RPC/FORK_URL --private-key $YOUR_ANVIL_PRIVATE_KEY --broadcast
```

That's why it's showing 0x00000.... ContractAddress found in broadcast latest. 

```stack
│ ├─ [0] VM::ffi([bash, /home/ecs/solidity-course/foundry-f23/foundry-fund-me-f23/lib/foundry-devops/src/get_recent_deployment.sh, FundMe, 31337, /home/ecs/solidity-course/foundry-f23/foundry-fund-me-f23//broadcast])
│ │ └─ ← 👎 0x0000000000000000000000000000000000000000
│ ├─ [0] console::log(Return Data:) [staticcall]
│ │ └─ ← ()
│ ├─ [0]
```

Because... Never broadcasted in both cases.
